pkg/util/pki/renewaltime.go

```diff
@@ -57,3 +52,43 @@ func RenewalTime(notBefore, notAfter time.Time, renewBeforeOverride *metav1.Dura
 	rt := metav1.NewTime(notAfter.Add(-1 * renewBefore).Truncate(time.Second))
 	return &rt
 }
+
+func RenewBefore(actualDuration time.Duration, renewBeforeOverride *string) time.Duration {
+	defaultRenewBefore := actualDuration / 3
+
+	if renewBeforeOverride == nil {
+		return defaultRenewBefore
+	}
+
+	d, pct, err := ParseDurationOrPercent(*renewBeforeOverride)
+	if err != nil {
+		return defaultRenewBefore
+	}
+
+	// If spec.renewBefore was set (and is less than duration)
+	// respect that. We don't want to prevent users from renewing
+	// longer lived certs more frequently.
+	if d != nil && *d > 0 && *d < actualDuration {
+		return *d
+	} else if pct != nil && *pct > 0 && *pct < 100 {
+		return actualDuration * time.Duration(100-*pct) / 100
+	}
+
+	return defaultRenewBefore
+}
+
+func ParseDurationOrPercent(s string) (*time.Duration, *int, error) {
+	if strings.HasSuffix(s, "%") {
+		v, err := strconv.Atoi(strings.TrimSuffix(s, "%"))
```

Comment:
For eg. CPU resource units, `0.5` is used instead of `50%`: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu
I wonder what the best format is here.

Reply:
I found an example of a percentage being used: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#rollingupdatedaemonset-v1-apps

Reply:
Yeah, the rolling update maxSurge/maxUnavailable fields were my inspiration here.

The IntOrString type those fields use made me consider a DurationOrPercent union type for this use case, but it ultimately felt unnecessary since both a duration and a percentage are serialized as strings in JSON (unlike int/string).